src/oss/langchain/streaming.mdx

```diff
@@ -509,6 +511,535 @@ for await (const [streamMode, chunk] of await agent.stream(
 ```
 :::
 
+:::python
+## Common patterns
+
+Below we include short, end-to-end snippets representing common use-cases for streaming.
+
+### Streaming tool calls
+
+You may want to stream both:
+
+1. Partial JSON as [tool calls](/oss/langchain/models#tool-calling) are generated
+2. The completed, parsed tool calls that are executed
+
+Specifying [`stream_mode="messages"`](#llm-tokens) will stream incremental [message chunks](/oss/langchain/messages#streaming-and-chunks) generated by all LLM calls in the agent. To access the completed messages with parsed tool calls:
+
+1. If those messages are tracked in the [state](/oss/langchain/agents#memory) (as in the model node of [`create_agent`](/oss/langchain/agents)), use `stream_mode=["messages", "updates"]` to access completed messages through [state updates](#agent-progress).
+2. If those messages are not tracked in the state, use [custom updates](#custom-updates) or aggregate the chunks during the streaming loop.
+
+Below, we demonstrate (1). We also provide an [end-to-end example of (2)](#streaming-completed-messages-that-are-not-in-agent-state) below.
```

Comment:
It would be better to put the code examples under H4s. Then from the numbered list above, you can add a "For an example of this, see x." and link to the section